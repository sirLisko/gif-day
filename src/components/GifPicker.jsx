import React, { Fragment, useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled/macro';

import theme from 'styles/theme';
import Modal from 'components/Modal';
import GifTile from 'components/GifTile';

const { REACT_APP_GIF_API_GET_RANDOM } = process.env;

const StyledContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
`;

const StyledInput = styled.input`
  border: 0;
  border-bottom: 1px solid ${theme.colors.accent};
  font-size: 2rem;
  outline-color: ${theme.colors.accent};
  padding: 0.25rem 1rem;
  margin: 1rem auto;
  width: calc(100% - 3.5rem);
`;

const StyledSearch = styled.button`
  background: ${theme.colors.accent};
  font-size: 1.5rem;
  border: navajowhite;
  border-radius: 50%;
  color: white;
  font-weight: bold;
  height: 3rem;
  width: 3rem;
  margin-left: 0.5rem;
  text-align: center;
  padding: 0;
`;

const StyledOk = styled.button`
  background: ${theme.colors.accent};
  background: linear-gradient(
    to right,
    ${theme.colors.accent} 0%,
    ${theme.colors.alternate} 80%,
    ${theme.colors.alternate} 100%
  );
  color: #fff;
  padding: 1rem 3rem;
  text-transform: uppercase;
  font-size: 1rem;
  font-weight: bold;
  border: none;
  border-radius: 5rem;
  margin: 1rem auto;
  display: block;
  letter-spacing: 3px;
  &:hover {
    transform: scale(1.04);
  }
`;

const GifPicker = ({ selectedDay, selectedImg, onGifSelected }) => {
  if (!selectedDay) {
    return null;
  }

  const [image, setImage] = useState(selectedImg);
  const [error, setError] = useState();
  const getGif = text =>
    fetch(`${REACT_APP_GIF_API_GET_RANDOM}${text}`)
      .then(response => response.json())
      .then(({ data }) => {
        if (Array.isArray(data)) {
          setError("We didin't find your Gif");
        } else {
          setError();
          setImage({ text, src: data.images.downsized_small.mp4 });
        }
      })
      .catch(() => setError('Please retry later.'));

  const textInput = useRef();
  useEffect(() => {
    textInput && textInput.current.focus();
  }, []);

  return (
    <Modal
      isModalOpen={Boolean(selectedDay)}
      onClose={() => onGifSelected(selectedImg)}
    >
      <StyledContainer>
        <form
          onSubmit={e => {
            e.preventDefault();
            return getGif(e.target.elements.what.value);
          }}
        >
          <StyledInput
            type="text"
            name="what"
            ref={textInput}
            defaultValue={image && image.text}
          />
          <StyledSearch type="submit">yo!</StyledSearch>
        </form>
        {error && (
          <div style={{ textAlign: 'center' }}>
            Ooops! Something went wrong :( <p>{error}</p>
          </div>
        )}
        {image && (
          <Fragment>
            <GifTile gif={image.src} />
            <StyledOk onClick={() => onGifSelected(image)}>
              You Got It!
            </StyledOk>
          </Fragment>
        )}
      </StyledContainer>
    </Modal>
  );
};

GifPicker.propTypes = {
  selectedDay: PropTypes.string.isRequired,
  selectedImg: PropTypes.shape({
    text: PropTypes.string.isRequired,
    src: PropTypes.string.isRequired,
  }),
  onGifSelected: PropTypes.func.isRequired,
};

export default GifPicker;
