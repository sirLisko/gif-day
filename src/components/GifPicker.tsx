import React, { Fragment, useState, useEffect, useRef } from "react";

import theme from "styles/theme";
import Modal from "components/Modal";
import GifTile from "components/GifTile";

import { getRandomGif } from "lib/gifAPI";
import { Image } from "types";

interface Props {
  selectedDay?: string;
  selectedImg?: Image;
  onClosePicker: (image?: Image) => void;
}

const GifPicker = ({ selectedDay, selectedImg, onClosePicker }: Props) => {
  const [image, setImage] = useState(selectedImg);
  const [error, setError] = useState<string>();
  const [loading, setLoading] = useState<boolean>();
  const getGif = (text: string) => {
    setLoading(true);
    fetch(`/api/search/${text}`)
      .then((response) => response.json())
      .then((gif) => {
        if (!gif) {
          return setError("we didn't find your gif");
        }
        setError(undefined);
        setImage({ text, gif });
        setLoading(false);
      })
      .catch(() => {
        setError("Please try again later");
        setLoading(false);
      });
  };

  const textInput = useRef<HTMLInputElement>(null);
  useEffect(() => {
    textInput && textInput.current?.focus();
  }, []);

  if (!selectedDay) {
    return null;
  }

  return (
    <Modal
      isModalOpen={Boolean(selectedDay)}
      onClose={() => onClosePicker(selectedImg)}
    >
      <div className="container">
        <form
          onSubmit={(e) => {
            e.preventDefault();
            return textInput.current?.value && getGif(textInput.current.value);
          }}
        >
          <input
            type="text"
            name="what"
            ref={textInput}
            defaultValue={image && image.text}
          />
          <button type="submit">yo!</button>
        </form>
        {error && (
          <div style={{ textAlign: "center" }}>
            Ooops! Something went wrong :( <p>{error}</p>
          </div>
        )}
        {loading && <span>loading...</span>}
        {image && (
          <Fragment>
            {!loading && (
              <Fragment>
                <GifTile gifObj={image} />
                <button onClick={() => onClosePicker(image)}>
                  You Got It!
                </button>
              </Fragment>
            )}
          </Fragment>
        )}
        <style jsx>{`
          .container {
            display: flex;
            flex-direction: column;
            justify-content: flex-start;
            align-items: center;
          }
          input {
            border: 0;
            border-bottom: 1px solid ${theme.colors.accent};
            font-size: 2rem;
            outline-color: ${theme.colors.accent};
            padding: 0.25rem 1rem;
            margin: 1rem auto;
            width: calc(100% - 3.5rem);
          }
          button[type="submit"] {
            border: 0;
            border-bottom: 1px solid ${theme.colors.accent};
            font-size: 2rem;
            outline-color: ${theme.colors.accent};
            padding: 0.25rem 1rem;
            margin: 1rem auto;
            width: calc(100% - 3.5rem);
          }
          button:not([type="submit"]) {
            background: ${theme.colors.accent};
            background: linear-gradient(
              to right,
              ${theme.colors.accent} 0%,
              ${theme.colors.alternate} 80%,
              ${theme.colors.alternate} 100%
            );
            color: #fff;
            padding: 1rem 3rem;
            text-transform: uppercase;
            font-size: 1rem;
            font-weight: bold;
            border: none;
            border-radius: 5rem;
            margin: 1rem auto;
            display: block;
            letter-spacing: 3px;
            &:hover {
              transform: scale(1.04);
            }
          }
        `}</style>
      </div>
    </Modal>
  );
};

export default GifPicker;
