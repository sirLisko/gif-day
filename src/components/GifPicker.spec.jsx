import React from 'react';
import { shallow } from 'enzyme';

import GifPicker from './GifPicker';

describe('GifPicker Component', () => {
  let wrapper;
  const props = {
    selectedDay: '0-0',
    onGifSelected: jest.fn(),
  };

  beforeEach(() => {
    fetch.resetMocks();
    wrapper = shallow(
      <GifPicker {...props}>
        <div>foo</div>
      </GifPicker>,
    );
  });

  it('should NOT render if no selectedDay', () => {
    wrapper.setProps({ selectedDay: null });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should fetch a new gif if form submitted', done => {
    const fakeEvent = {
      preventDefault: jest.fn(),
      target: {
        elements: {
          what: {
            value: 'foo',
          },
        },
      },
    };
    const fakeResponse = {
      data: { images: { downsized_small: { mp4: 'foobar' } } },
    };
    wrapper = shallow(
      <GifPicker {...props}>
        <div>foo</div>
      </GifPicker>,
    );
    fetch.mockResponseOnce(JSON.stringify(fakeResponse));
    wrapper.find('form').simulate('submit', fakeEvent);
    expect(wrapper.find('video').length).toBe(0);
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });
});
