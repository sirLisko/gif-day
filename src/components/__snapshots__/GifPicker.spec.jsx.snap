// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifPicker Component should NOT render if no selectedDay 1`] = `""`;

exports[`GifPicker Component should fetch a new gif if form submitted 1`] = `
<Modal
  isModalOpen={true}
  onClose={[Function]}
>
  <StyledContainer>
    <form
      onSubmit={[Function]}
    >
      <StyledInput
        defaultValue="foo"
        name="what"
        type="text"
      />
      <StyledSearch
        type="submit"
      >
        yo!
      </StyledSearch>
    </form>
    <GifTile
      gif="foobar"
    />
    <StyledOk
      onClick={[Function]}
    >
      You Got It!
    </StyledOk>
  </StyledContainer>
</Modal>
`;

exports[`GifPicker Component should render properly 1`] = `
<Modal
  isModalOpen={true}
  onClose={[Function]}
>
  <StyledContainer>
    <form
      onSubmit={[Function]}
    >
      <StyledInput
        name="what"
        type="text"
      />
      <StyledSearch
        type="submit"
      >
        yo!
      </StyledSearch>
    </form>
  </StyledContainer>
</Modal>
`;
